File: structy/02-linked-list/07-zipper-lists/02.js
/*
Complexity
- Time: O(min(m, n))
- Space: O(n)

*/

import Node from '../node.js';

const zipperLists = (head1, head2) => {
    if (head1 === null && head2 === null) return null; 
    if (head1 === null) return head2;
    if (head2 === null) return head1;

    const next1 = head1.next;
    const next2 = head2.next;
    head1.next = head2;
    head2.next = zipperLists(next1, next2);
    return head1;
};



const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
a.next = b;
b.next = c;
// a -> b -> c

const x = new Node("x");
const y = new Node("y");
const z = new Node("z");
x.next = y;
y.next = z;

// x -> y -> z

console.log(zipperLists(a, x)); // a -> x -> b -> y -> c -> z

File: structy/02-linked-list/03-linked-list-find/02.py
# linked list find
# Write a function, linked_list_find, that takes in the head of a linked list and a target value. The function should return a boolean indicating whether or not the linked list contains the target.

"""
INPUT: 
- head: Node
- target: Any

OUTPUT: 
- boolean

RULES 
- it should return a boolean if the target value is found in the list. Otherwise, return false

ALGORITHM (Iterative)
----------

FUNC linked_list_find(head: Node, target: Any) -> Bool
- WHILE the current node is not NULL 
    - IF the current node has the target value 
        - RETURN TRUE
- RETURN FALSE 
ENDFUNC

COMPLEXITY
- N: number of nodes
- Time: O(N)
- Space: O(1)

ALGORITHM (recursive)
---------

FUNC linked_list_find(head: Node, target: Any) -> Bool
- IF current node is None
    - RETURN FALSE
- IF current node has the target value 
    - RETURN TRUE
- RETURN recursive call to `linked_list_find` with the next node and the target value
ENDFUNC

COMPLEXITY
N: number of nodes in LL
Time: O(N)
Space: O(N)
"""

from typing import Any

class Node: 
    def __init__(self, val) -> None:
        self.val = val
        self.next = None

def linked_list_find(head: Node, target: Any):
    if head is None: 
        return False
    if head.val == target:
        return True
    return linked_list_find(head.next, target)

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d
print(linked_list_find(a, "c")) # True




a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d
print(linked_list_find(a, "q")) # False




node1 = Node("jason")
node2 = Node("leneli")

node1.next = node2

# jason -> leneli
print(linked_list_find(node1, "jason")) # True
